src/typing/flow_js.ml

```diff
@@ -4883,22 +4889,60 @@ let rec __flow cx ((l: Type.t), (u: Type.use_t)) trace =
       then (
         structural_subtype cx trace ~use_op l reason_inst
           (fields_tmap, methods_tmap);
+        let reason = replace_reason (fun desc -> RStatics desc) reason in
+        let fn = get_builtin_type cx ~trace reason "Function" in
+        structural_subtype cx trace ~use_op fn reason_stat
+          (static_fields_tmap, static_methods_tmap);
         rec_flow cx trace (l, UseT (use_op, super))
       )
 
     (***************************************************************)
     (* Enable structural subtyping for upperbounds like interfaces *)
     (***************************************************************)
 
-    | _,
-      UseT (use_op, DefT (reason_inst, InstanceT (_, super, _, {
-        fields_tmap;
-        methods_tmap;
-        structural = true;
-        _;
+    | DefT (lr, (VoidT | NullT)),
+      UseT (_, DefT (ur, InstanceT (_, _, _, { structural = true; _ }))) ->
+        (* `undefined` and `null` lack a static `name` property. *)
+        add_output cx ~trace (FlowError.EInterfaceIncompatibility (lr, ur))
+
+    | _, UseT (use_op, DefT (reason_inst, InstanceT (
+        DefT (reason_stat, InstanceT (_, _, _, {
+          fields_tmap = static_fields_tmap;
+          methods_tmap = static_methods_tmap;
+          structural = true;
+          _;
+        })), super, _, {
+          fields_tmap;
+          methods_tmap;
+          structural = true;
+          _;
       }))) ->
       structural_subtype cx trace ~use_op l reason_inst
         (fields_tmap, methods_tmap);
+      let l_static = match l with
+        | DefT (reason, ObjT _) ->
+            (* This would throw away any statics on those lower bounds
+               constructed through `FunT _ ~> ConstructorT _` if the statics
+               hadn't already been thrown away. *)
+            let reason = replace_reason (fun desc -> RStatics desc) reason in
+            let obj = get_builtin_type cx ~trace reason "Object" in
+            lookup_static cx trace reason obj
+        | DefT (reason, ArrT arrtype) ->
+            let array = get_array cx ~trace reason arrtype in
+            lookup_static cx trace reason array
+        | DefT (reason, StrT _) ->
+            let string = get_builtin_type cx ~trace reason "String" in
+            lookup_static cx trace reason string
+        | DefT (reason, BoolT _) ->
+            let boolean = get_builtin_type cx ~trace reason "Boolean" in
+            lookup_static cx trace reason boolean
+        | DefT (reason, NumT _) ->
+            let number = get_builtin_type cx ~trace reason "Number" in
+            lookup_static cx trace reason number
+        | _ -> assert false
```

Comment:
I'm prefer failing loudly on unhandled edge cases, but this may be a little too loud, no? Is there some extant error that invites the user to open an issue?